### ISSUE TITLE
Run applications in a JPMS way

### ISSUE THREAD

**Issue by Thihup** [piranhacloud/piranha#1305]
This is mostly an experimental thing. Current the Servlet (and any other) spec doesn't define how a container should handle a JPMS application.

I'd like to see if maybe we could get something workable and maybe help the specs in how the behavior should be.

I have been playing around and the java.lang.module API is very complex. The API contains the class ModuleFinder who can find the modules, but it only handles Paths ([ModuleFinder.html#of(java.nio.file.Path...)](https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/module/ModuleFinder.html#of(java.nio.file.Path...)), making it filesystem dependant.

Ideally, we should use this API because it's implementation does exactly what we are used to. However, the can create our own ModuleFinder that doesn't depend on the JDK implementation. This way, we can create a ModuleFinder for "wars", and define our own concept of some features, like "patch-module" and were the module-info will be located (in the root of the war, in the WEB-INF/classes/module-info) and so on.

In this case, it would be implemented using layers, so an application server could implement it even it's underlying runtime doesn't use JPMS.

Sure that this is a long topic. We may even propose another file format, like "mar" to be a modular war to not be confused with the current behavior of the wars.

Comments?

**Comment by manorrock** [comment]
In my mind we should be able to give the WebApplication a module path to load JARs from and yes that should come from both the filesystem and/or using something like we have in place for Piranha Micro, but then for module JARs ;)

**Comment by Thihup** [comment]
Probably we can use something from here:
http://mail.openjdk.java.net/pipermail/jigsaw-dev/2018-September/013924.html

**Comment by Thihup** [comment]
Happy new year!

I've got something working locally and it can deploy all the Servlet TCK wars in a new layer for each application. I had to work around some behavior of the classloader provided by the ModuleLayer, but as far as I can tell it is working nicely.

It first tries to load the module-info from the resource, then it tries to load the "Automatic-Module-Name" from the manifest, and then it derives the module name from the resource name.

If there are any split package (as some TCK tests have) or any other issues creating a module, it fallbacks to the unnamed module and works as usual.

I need to clean up some code and add more tests, but it is looking promising ;)
